Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/KeyFileDownloader.kt

```diff
@@ -0,0 +1,303 @@
+                    path
+                }
+            }
+            .also { Timber.tag(TAG).d("Returning %d available keyfiles", it.size) }
+    }
+
+    /**
+     * Fetches files given by serverDates by respecting countries
+     * @param availableCountries pair of dates per country code
+     */
+    @Suppress("LongMethod")
+    private suspend fun syncMissingDays(
+        availableCountries: List<LocationCode>
+    ) = withContext(Dispatchers.IO) {
+        val availableCountriesWithDays = availableCountries.map {
+            val days = diagnosisKeyServer.getDayIndex(it)
+            CountryDays(it, days)
+        }
+
+        val cachedDays = keyCache
+            .getEntriesForType(CachedKeyInfo.Type.COUNTRY_DAY)
+            .filter { it.first.isDownloadComplete && it.second.exists() } // We overwrite not completed ones
+            .map { it.first }
+
+        // All cached files that are no longer on the server are considered stale
+        val staleKeyFiles = cachedDays.filter { cachedKeyFile ->
+            val availableCountry = availableCountriesWithDays.singleOrNull {
+                it.country == cachedKeyFile.location
+            }
+            if (availableCountry == null) {
+                Timber.tag(TAG).w(
+                    "Unknown location %s, assuming stale cache.",
+                    cachedKeyFile.location
+                )
+                return@filter true // It's stale
+            }
+
+            availableCountry.dayData.none { date ->
+                cachedKeyFile.day == date
+            }
+        }
+        if (staleKeyFiles.isNotEmpty()) keyCache.delete(staleKeyFiles)
+
+        val nonStaleDays = cachedDays.minus(staleKeyFiles)
+        val countriesWithMissingDays = availableCountriesWithDays.mapNotNull {
+            it.toMissingDays(nonStaleDays)
+        }
+        Timber.tag(TAG).d("Downloading missing days: %s", countriesWithMissingDays)
+        val batchDownloadStart = System.currentTimeMillis()
+        val dayDownloads = countriesWithMissingDays
+            .flatMap { country ->
+                country.dayData.map { dayDate -> country to dayDate }
+            }
+            .map { (countryWrapper, dayDate) ->
+                async {
+                    val (keyInfo, path) = keyCache.createCacheEntry(
+                        location = countryWrapper.country,
+                        dayIdentifier = dayDate,
+                        hourIdentifier = null,
+                        type = CachedKeyInfo.Type.COUNTRY_DAY
+                    )
+
+                    return@async try {
+                        downloadKeyFile(keyInfo, path)
+                        keyInfo to path
+                    } catch (e: Exception) {
+                        Timber.tag(TAG).e(e, "Download failed: %s", keyInfo)
+                        null
+                    }
+                }
+            }
+
+        Timber.tag(TAG).d("Waiting for %d missing day downloads.", dayDownloads.size)
```

Comment:
AFAICT no. This will only be called after `dayDownloads` is completely mapped, so `dayDownloads` contains the correct amount of `Deferred<...>` items.